frontends/web/src/components/dialog/dialog.tsx

```diff
@@ -0,0 +1,135 @@
+/**
+ * Copyright 2018 Shift Devices AG
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Component, h, RenderableProps } from 'preact';
+import * as style from './dialog.css';
+
+interface Props {
+    title?: string;
+    small?: boolean;
+    disableEscape?: boolean;
+    onClose: (e: Event) => void;
+}
+
+interface State {
+    active: boolean;
+    currentTab: number;
+}
+
+class Dialog extends Component<Props, State> {
+    private modalContent!: HTMLElement;
+    private focusableChildren!: NodeListOf<HTMLElement>;
+
+    constructor(props: Props) {
+        super(props);
+        this.state = {
+            active: false,
+            currentTab: 0,
+        };
+    }
+
+    public componentDidMount() {
+        setTimeout(this.activate, 10);
+    }
+
+    public componentWillUnmount() {
+        this.deactivate();
+    }
+
+    private setModalContent = (element: HTMLElement) => {
+        this.modalContent = element;
+    }
+
+    private focusWithin = () => {
+        this.focusableChildren = this.modalContent.querySelectorAll('a, button, input, textarea');
+        for (const c of this.focusableChildren) {
+            c.classList.add('tabbable');
+        }
+        document.addEventListener('keydown', this.controlKeys);
+    }
+
+    private addTabIndex = (currentTab: number) => {
+        let next = currentTab + 1;
+        if (next >= this.focusableChildren.length) {
+            next = 0;
+        }
+        this.setState({ currentTab: next });
```

Comment:
You could shrink this method to one line (but I don't care if you want to do it this time):
```
this.setState({ currentTab: (currentTab + 1) % this.focusableChildren.length });
```